### PR TITLE
Compile all system libraries with `-g`

### PR DESCRIPTION
**Pull request by sbc100**
As an alternative to shipping unoptimized debug copies of all
libraries just add debug into the optimized libraries.  Release
builds are linked with `--strip-debug` so should be (mostly) unaffected.

Fixes: #12060, #9703, #13131

### CODE REVIEW

**Comment by bmeurer** [comment]
This is going to make the link step slow, even if debug information is not requested. It also blows up the resulting binary since it's not using split DWARF (which arguably doesn't work yet, but should soon).

**Comment by sbc100** [comment]
> This is going to make the link step slow, even if debug information is not requested. 

Why makes you think this will be the case?  I would have though that passing `--strip-debug` to the linker (which emcc does in release mode) would make the code of the extra info here very minimal (just the cost a little more mmap space). 

I'm going to do some experiments with release builds of both small and large (ish) apps to see if I can measure the difference.   For large apps I imagine it will not be relevant since the standard libraries will be dwarf'd (no pun intended) by the application code.

**Comment by bmeurer** [comment]
> > This is going to make the link step slow, even if debug information is not requested.
> 
> Why makes you think this will be the case? I would have though that passing `--strip-debug` to the linker (which emcc does in release mode) would make the code of the extra info here very minimal (just the cost a little more mmap space).
> 
> I'm going to do some experiments with release builds of both small and large (ish) apps to see if I can measure the difference. For large apps I imagine it will not be relevant since the standard libraries will be dwarf'd (no pun intended) by the application code.

I think the way it works is that the debug sections are stripped after the final link step (I might be mistaken here).

The other problem remains still, in that this will not work with split DWARF, once we get it to work, all debug information from the system libraries will end up in the final binary (or the `-gseparate-dwarf` file).

**Comment by sbc100** [comment]
> > > This is going to make the link step slow, even if debug information is not requested.
> > 
> > 
> > Why makes you think this will be the case? I would have though that passing `--strip-debug` to the linker (which emcc does in release mode) would make the code of the extra info here very minimal (just the cost a little more mmap space).
> > I'm going to do some experiments with release builds of both small and large (ish) apps to see if I can measure the difference. For large apps I imagine it will not be relevant since the standard libraries will be dwarf'd (no pun intended) by the application code.
> 
> I think the way it works is that the debug sections are stripped after the final link step (I might be mistaken here).

The debug into is stripped when wasm-ld read is (because we pass `--strip-debug` to the linker).

See: https://github.com/emscripten-core/emscripten/blob/2b9d22d7fdf13b239235052b55811eb24fe5e2b5/tools/building.py#L496
And: https://github.com/llvm/llvm-project/blob/c4fc8a21d1dbfd2e269a59533808913b46faa8ce/lld/wasm/Writer.cpp#L139

> 
> The other problem remains still, in that this will not work with split DWARF, once we get it to work, all debug information from the system libraries will end up in the final binary (or the `-gseparate-dwarf` file).

Well I guess we will indeed need a separate set of libraries to make that use case work.   But I think we can consider that issue separately.   At least this change allows folks using regular `-g` builds to get this working without the extra library set.  So rather than shipping 3 sets (1 release, 2 debug, 2 split-debug) we only need t 2 sets (1 debug, 2 split-debug)

**Comment by bmeurer** [comment]
Ack, sounds reasonable. 👍

File: tools/system_libs.py [hunk 1]
```diff
@@ -55,8 +55,10 @@ def dir_is_newer(dir_a, dir_b):
   return newest_a < newest_b
 
 
-def get_cflags(force_object_files=False):
-  flags = []
+def get_base_cflags(force_object_files=False):
+  # Always build system libraries with debug information.  Non-debug builds will ignore this
+  # at link time because then link with `-strip-debug`.
```
Comment (kripken) [inline tools/system_libs.py:60]:
```suggestion
  # at link time because we then link with `-strip-debug`.
```

File: tests/emscripten_log/emscripten_log.cpp [hunk 2]
```diff
@@ -80,7 +80,12 @@ void __attribute__((noinline)) bar(int = 0, char * = 0, double = 0) {
     but the line numbers will greatly vary depending on the mode we are compiling in, so cannot test with direct string comparison. */
 
   if ((flags & EM_LOG_C_STACK) != 0) {
+    // TODO(https://github.com/emscripten-core/emscripten/issues/13089)
+    // We should be able to check for emscripten_log.cpp here but sadly
```
Comment (kripken) [inline tests/emscripten_log/emscripten_log.cpp:84]: from the linked bug, this was always an issue, and not uncovered by this PR's adding `-g` to libraries? if so could it be a separate PR? I'm not actually suggesting it be, this looks fine, but I'd like to understand if this has to be here or not.

**Review by kripken — APPROVED** [review]
lgtm, assuming

[ ] you've confirmed that link times are not affected in non-debug builds
[ ] that the regression in link times for debug builds is reasonable. i wonder in particular if a test suite with `-g` in it, like `wasm2g`, becomes much slower on our CI with this PR
[ ] you've checked how much larger this makes the emsdk downloads, and it's not a big deal

File: tests/emscripten_log/emscripten_log.cpp (hunk 2 above)
Comment (sbc100) [reply tests/emscripten_log/emscripten_log.cpp:84]: The bug already existed but this PR makes it show up.   The problem is that the source map seems to be re-read and when there is only on file in it it shows `emscripten_log.cpp` but when there are other files in the source map it shows them (and now `emscripten_log.cpp`) in a way that I don't understand, and is not correct either before or after this change.

**Comment by sbc100** [comment]
non-debug link times
----------------------------

I measured non-debug build and debug build time with `-s MAIN_MODULE` (which includes the entire standard library, so is maximally slow in terms of stdlib link time).

Before: non-debug: 1.2-1.2s  debug: 1.27-1.35
After:  non-debug: 1.1-1.2s  debug: 2.0-2.1s

So as expected non-debug link time is not measurably effected.    debug link times are (as expected)  effected and go up by about 40% for hello world, but this increase of (0.6s on my machine) is likely to be fixed an will not scale with program size so will not be relevant to large programs.

effect on test suite
------------------------

running the entire wasm2g test suite (with EMTEST_SKIP_SLOW to skip the really slow tests):
Before: 61.9, 56.0, 55.9, 56.1 (mean: 57.4)
After: 61.7, 62.2, 62.5, 62.6 (mean: 62.0)
difference 8%

So it does effect the time to run debug tests, but that is completely expected, debug tests are slow for exactly this kind of reason.

emsdk download size
-----------------------------

Cache directory size (after running `./embuilder build SYSTEM`)
Before:  33Mb (zipped: 6Mb)
After:  73Mb  (zipped: 21Mb) +120% (zipped +500%)

The current emsdk download is 241Mb zipped, with the additional 15Mb of zipped debug info would be a 6% increase in total download.   Unzipped the current SDK is 848 so the extra 50Mb would be a 5% increase in disc space.

**Comment by sbc100** [comment]
We don't yet have agreement if whether this is something we want to enable by default.   The present out there from linux distros and from windows is not include debuggable system libraries by default but to have some kind of opt-in mechanism for downloading or enabling them.   Do we have a strong enough case to diverge here?   Do our users want to step into system libraries often enough?

**Comment by kripken** [comment]
I see I approved this a while ago, but reading it now I'm not 100% sure I should have, heh...

At minimum how about adding a changelog entry for this, and seeing if there are any regressions reported?

**Comment by sbc100** [comment]
OK I'll add a changelog.

Are worried about the size regression in the emsdk download?   I don't see any other downsides, but I might be missing something.

**Comment by sbc100** [comment]
BTW, I manually verified that std::string can be effectively views while debugging in devtools after this change.

**Comment by kripken** [comment]
Oh, nice about `std::string`! Do you mean the values, or stack traces, or both?

**Comment by sbc100** [comment]
The values in in the local inspector.  Without this change I just saw "undefined".

**Comment by kripken** [comment]
Excellent!
